### PR TITLE
Feature/limitcheck

### LINKED ISSUE #46: setValue does not check min/max
Currently, I have the following observation: 

```
VSS Client> getMetaData Vehicle.Speed
{
    "action": "getMetaData",
    "metadata": {
        "Vehicle": {
            "children": {
                "Speed": {
                    "datatype": "int32",
                    "description": "Vehicle speed, as sensed by the gearbox.",
                    "max": 250,
                    "min": -250,
                    "type": "sensor",
                    "unit": "km/h",
                    "uuid": "1efc9a11943b5a15ac159786051c5836"
                }
            },
            "description": "High-level vehicle data.",
            "type": "branch",
            "uuid": "1c72453e738511e9b29ad46a6a4b77e9"
        }
    },
    "requestId": "1236",
    "timestamp": 1599086861
}
VSS Client> setValue Vehicle.Speed 366
{
    "action": "set",
    "requestId": "1235",
    "timestamp": 1599087046
}
VSS Client> getValue Vehicle.Speed 
{
    "action": "get",
    "path": "Vehicle.Speed",
    "requestId": "1234",
    "timestamp": 1599087196,
    "value": "366"
}
```
But I do not expect the `setValue  Vehicle.Speed 366` can be successful, because the defined max value is 250.

### PR DESCRIPTION
**Pull request by SebastianSchildt**
Fixes #46 and rewrites the type check code

### CODE REVIEW

File: src/VSSTypeSanitizer.cpp [hunk 1]
```diff
@@ -0,0 +1,114 @@
...
+        throw outOfBoundException(msg.str());
+    }
+}
+
+
+/** This will check whether &val val is a valid value for the sensor described
+ *  by meta  and whether  it is within the limits defined by VSS if any */
+void VssDatabase::checkAndSanitizeType(jsoncons::json &meta, jsoncons::json &val) {
+    std::string dt=meta["datatype"].as<std::string>();
+    if (dt == "uint8") {
```
Comment (wenwenchenbosch) [inline src/VSSTypeSanitizer.cpp:73]: so many ifs... but it is ok :)

File: src/VSSTypeSanitizer.cpp [hunk 2]
```diff
@@ -0,0 +1,114 @@
...
+
+
+
+
+
+template<typename T>
+void checkNumTypes(jsoncons::json &meta, jsoncons::json &val )
+{
+    //Double can be larger than largest allowed VSS type ((u)int64)
+    double dval = val.as<long double>();
```
Comment (wenwenchenbosch) [inline src/VSSTypeSanitizer.cpp:30]: use https://en.cppreference.com/w/cpp/string/basic_string/stol or maybe there is better function for it. I did not search so much. but what you did is unsafe. see the failing testcases :p

**Review by wenwenchenbosch — CHANGES_REQUESTED** [review]

**Comment by wenwenchenbosch** [comment]
now it is broken: https://kuksaval.northeurope.cloudapp.azure.com/blue/organizations/jenkins/kuksaval-upstream/detail/PR-148/2/tests

File: test/unit-test/KuksavalUnitTest.cpp [hunk 3]
```diff
@@ -135,13 +135,13 @@ BOOST_AUTO_TEST_CASE(set_get_test_all_datatypes_boundary_conditions)
     string test_path_Uint32 = "Vehicle.OBD.TimeSinceDTCCleared";
     VSSPath vss_test_path_Uint32 = VSSPath::fromVSSGen1(test_path_Uint32);
 
-    string test_path_int8 = "Vehicle.OBD.ShortTermFuelTrim2";
+    string test_path_int8 = "Vehicle.Body.Mirrors.Right.Tilt";
```
[the remark this answers is not in the document]
Comment (SebastianSchildt) [reply test/unit-test/KuksavalUnitTest.cpp:138]: test requires unlimited sensors (without min/max)

**Review by wenwenchenbosch — APPROVED** [review]

**Pull request merged**